// @hybris-software/use-table/dist/src/TableClient/HeaderActionList/HeaderActionList.js
"use strict";

require("core-js/modules/es.weak-map.js");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("core-js/modules/web.dom-collections.iterator.js");
var _react = _interopRequireWildcard(require("react"));
var _ConditionalComponent = _interopRequireDefault(require("../ConditionalComponent/ConditionalComponent"));
var _bi = require("react-icons/bi");
var _HeaderActionListModule = _interopRequireDefault(require("./HeaderActionList.module.css"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
//Components

//Icon

// Styles

function HeaderActionList(_ref) {
  let {
    column,
    tableRef,
    setHiddenColumns,
    texts
  } = _ref;
  const [showActionList, setShowActionList] = (0, _react.useState)(false);
  return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_bi.BiDotsVerticalRounded, {
    onClick: () => setShowActionList(!showActionList),
    className: _HeaderActionListModule.default.headerActions,
    size: "20"
  }), /*#__PURE__*/_react.default.createElement(_ConditionalComponent.default, {
    condition: showActionList
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _HeaderActionListModule.default.boxAction,
    onMouseLeave: () => setShowActionList(false)
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _HeaderActionListModule.default.columnActions
  }, /*#__PURE__*/_react.default.createElement(_ConditionalComponent.default, {
    condition: column.sortable !== false
  }, /*#__PURE__*/_react.default.createElement("div", {
    onClick: () => {
      const columnName = column.field;
      tableRef.current.setSortingSettings(columnName);
    }
  }, texts.columnByAsc)), /*#__PURE__*/_react.default.createElement(_ConditionalComponent.default, {
    condition: column.sortable !== false
  }, /*#__PURE__*/_react.default.createElement("div", {
    onClick: () => {
      const columnName = column.field;
      tableRef.current.setSortingSettings("-" + columnName);
    }
  }, texts.columnByDesc)), /*#__PURE__*/_react.default.createElement("div", {
    onClick: () => {
      setHiddenColumns(oldState => [...oldState, column.field]);
    }
  }, texts.hideColumn), /*#__PURE__*/_react.default.createElement("div", {
    onClick: () => {
      setHiddenColumns([]);
    }
  }, texts.showColumns)))));
}
var _default = HeaderActionList;
exports.default = _default;
